### PR TITLE
Upgrade to support higher security hashes and deriving encrypt AND signing keys

### PR DESCRIPTION
**Pull request by grempe**
Hi, I needed pretty much what you had built (thanks) but I also wanted to generate high quality signing keys as well as encryption keys.

I have gone through the code, tests, and readme and made significant updates including the concept of `v1` and `v2` versions of the main function.  v1 is fully backward compatible and v2 provides higher security through as 64 byte BLAKE2b (vs. BLAKE2s) keyed hash and uses 64 bytes of scrypt output.  The 64 bytes of output is split in two to provide two different seeds, one for generating encryption keys, and one for signing keys.

I also return the Base64 version of the keys for convenience. The tests vectors for v1are unmodified from your original and I've added new tests for the v2 version which is now the default. A user upgrade from the old version should only require adding the 'v1' version argument to their existing code block.

I bumped the version of the package to 2.0.0 to fit with semver.

I hope you can pull this changeset upstream and push a new version.

Thanks!

Glenn

### CODE REVIEW

**Comment by jo** [comment]
Thank you @grempe for this epic and very much appreciated pull request! I have read through your changes carefully and absolutely agree with both the idea and your implementation. Much kudos for your documention improvements, or better say for making a documentation at all, since before there was barely.

Please give me one more night to think about the implications and your changes. There is so much to learn for me in there.

My first question is not really related to the merge of this pull: While reading your commits one by one I stumbled upon your [idea of using a hash of the private key for the signing key](https://github.com/jo/session25519/pull/2/commits/429bd628a8ef9e48768b9c56bcc2eacf014d5540#diff-04c6e90faac2675aa89e2176d2eec7d8R74) and immediately thought this might be insecure. And you also replaced that method by [generating a 64-bit private key and split it into two 32-bit private key seeds](https://github.com/jo/session25519/pull/2/commits/b7c96982fef4ff0cecc753f523bf81ec0ff92137#diff-04c6e90faac2675aa89e2176d2eec7d8R145) later. But why does this trigger my alarm bells? Would it be secure?

The second thing I found is that the size of the build `dist/session25519.js` has increased from   107 to 143 KB ([2,469 additions, 1,457 deletions](https://github.com/jo/session25519/pull/2/files#diff-b019c02c4170fe29fab1e7e6e1b764c3)). Not that its a problem just for me to know - is this because of [the replacement of `blake2s-js` with `blakejs`](https://github.com/jo/session25519/pull/2/commits/5c2223231a60c5b2b5419261a3f9f9ae1950d2cd)?

Thanks!

File: README.md [hunk 1]
```diff
@@ -1,43 +1,201 @@
...
+There are two versions of this function that you can call, `v1` and `v2`. `v1` is
+the original version (before there was a version). If you are upgrading from
+version `1.x.x` of this package you will want to specify the `v1` arg.  If you
+do not, the keys generated will not be the same deterministic keypair you would
+expect for a given user ID and password. If you specify `v1`, which is backwards
+compatible, everything should behave identically to before.
+
+The new `v2` version enhances the security of the system, and provides an
+additional key pair. The hash of the password is now a 64 Byte (vs. 32B in v1)
+`BLAKE2b` hash (vs. `BLAKE2s` in `v1`). This hash is alse keyed, like an HMAC,
```
Comment (jo) [inline README.md]: a typo?

``` diff
- This hash is alse keyed
+ This hash is also keyed
```

File: README.md [hunk 2]
```diff
@@ -1,43 +1,201 @@
...
+
+```sh
+npm run build
+```
 
 ## Test
+
 ```sh
-npm test
+npm run test
```
Comment (jo) [inline README.md]: Why this change?

**Comment by grempe** [comment]
Hi @jo, glad you are enjoying the pull request.  To answer your main question about the security of splitting the 64 Byte hash in two in order to use the two halves as seeds. I was implementing something along these lines a few months ago and I asked @dchest (author of tweetnacl.js) about the security of this and he responded in an email.

```
If your user has 32 secret random bytes, but you need 64 random bytes — 32 bytes for nacl.box secret key, and 32 bytes for nacl.sign secret key seed — the correct answer is not to look into the guts of algorithms to figure out if their relationship is safe, but to cryptographically expand those 32 bytes into 64 bytes. If you’re not planning to derive more keys from it in the future, a simple SHA512 will work fine:

  var secret = nacl.randomBytes(32); // this is what a user will store
  var expandedSecret = nacl.hash(secret);
  var boxKeyPair = nacl.box.keyPair.fromSecretKey(expandedSecret.subarray(0, 32));
  var signKeyPair = nacl.sign.keyPair.fromSeed(expandedSecret.subarray(32, 64)); 

By the way, you can have a single _keypair_ for both signing and boxing, by calculating a DH keypair from the signing one: see https://github.com/dchest/ed2curve-js (There’s no proof that this is safe yet, but it is considered safe by cryptographers.)
```

While writing this he just weighed in on Twitter saying the same:

https://twitter.com/dchest/status/711597559563862016

File: README.md (hunk 2 above)
Comment (grempe) [inline README.md]: Fixed in latest commits.

File: README.md (hunk 1 above)
Comment (grempe) [inline README.md]: This line was removed in updated README. Good catch though!

**Comment by grempe** [comment]
@jo here are some new commits on the pull request that reflect your questions and the discussion on Twitter today with @dchest

The code is now a bit simpler, using BLAKE2s 32B hash as the scrypt input for both v1 and v2 as @dchest thought the use of the 64 Byte hash didn't really offer that much security wise.

Regarding your file size question I suspect the increase is due to both the change in blake lib and in the use of the browserify --standalone flag which I needed to make it work in the browser for me.  I wasn't watching it too closely though.

I had previously bumped the version string up to 2.0, I'll leave it up to you if that is ok or if you want to use 2.1 for these new commits even though 2.0 was never released.

I added a shout-out to @dchest in the README.md as well. He deserves it. :-)

Cheers

**Comment by grempe** [comment]
Oh, and another reason for the small size increase in the dist file is the addition of the base64 lib.

**Comment by grempe** [comment]
I removed the dependency on the recently extracted nacl-util-js package.  I had reduced its usage to only a single call and I extracted the code from @dchest code and embedded it here to simplify.

Question. You have the .travis.yml file (which I updated) but you don't seem to be running it automatically on builds. If you are using that it would be good to push this through CI to make sure things are OK.

**Comment by jo** [comment]
I activated CI on Travis, thanks for pointing that out!

**Comment by grempe** [comment]
ok, all green now.  Unless you see other issues or enhancements I think it is good to go.  Good collaboration with @jo and @dchest.

**Comment by jo** [comment]
Good work! Will do final review tomorrow.

\o/

File: .travis.yml [hunk 3]
```diff
@@ -1,6 +1,20 @@
...
+  - "4"
+  - "4.2"
+  - "4.1"
+  - "4.0"
   - "0.12"
+  - "0.11"
+  - "0.10"
   - "iojs"
+env:
+  - CXX=g++-4.8
```
Comment (jo) [inline .travis.yml]: Is this still needed?

File: package.json [hunk 4]
```diff
@@ -1,11 +1,11 @@
 {
   "name": "session25519",
-  "version": "1.0.0",
-  "description": "Derive curve25519 key pair from username/password via scrypt.",
+  "version": "2.0.0",
```
Comment (jo) [inline package.json]: No need for breaking release, this is definitively `1.1.0` :100:

File: README.md [hunk 5]
```diff
@@ -1,43 +1,161 @@
...
+It bears repeating that **the strength of this system is very strongly
+tied to the strength of the passphrase chosen by the user**. Application
+developers are **strongly encouraged** to enforce the use of
+high-entropy pass phrases by users. Memorable high-entropy pass phrases,
+such as can be generated with [Diceware](https://www.rempe.us/diceware/),
+and measured with password strength estimation tools like
+[zxcvbn](https://github.com/dropbox/zxcvbn) are critically important to
+the overall security of the system.
+
+## Version 2.x.x Changes
```
Comment (jo) [inline README.md]:
``` diff
- ## Version 2.x.x Changes 
+ ## Version 1.1.x Changes 
```

**Comment by jo** [comment]
I made some comments, everything else looks very good to me.

Very good collaboration indeed!

**Comment by grempe** [comment]
Changes made.  All green.

**Comment by jo** [comment]
Awesome!

**Comment by jo** [comment]
```
[jo@linux session25519]$ npm publish
+ session25519@1.1.0
```

File: test.js [hunk 6]
```diff
@@ -1,17 +1,24 @@
...
+  session25519('user@example.com', 'secret', function (err, key) {
     t.notOk(err, 'no error')
-    
+
     t.deepEqual(key, {
-      publicKey: nacl.util.decodeBase64('EHfJ2sJJt7LjamnuJul6TlCb5SkgAwJI1PP7q95jh04='),
-      secretKey: nacl.util.decodeBase64('p6FU22u0423AcS0AKY6I+XeBTMDzQLX57XOH9Xg28pw=')
-    }, 'correct key has been calculated')
-    
+      publicKey: base64.toByteArray('EHfJ2sJJt7LjamnuJul6TlCb5SkgAwJI1PP7q95jh04='),
```
Comment (elimisteve) [inline test.js:12]: Same as before, yay! Therefore this appears backward-compatible with v1, and therefore theoretically compatible with miniLock, right @grempe?

File: test.js [hunk 7]
```diff
@@ -1,17 +1,24 @@
...
+      publicKeyBase64: 'EHfJ2sJJt7LjamnuJul6TlCb5SkgAwJI1PP7q95jh04=',
+      secretKey: base64.toByteArray('p6FU22u0423AcS0AKY6I+XeBTMDzQLX57XOH9Xg28pw='),
```
Comment (elimisteve) [inline test.js:14]: Also same :-D

File: index.js [hunk 8]
```diff
@@ -1,34 +1,82 @@
...
-    } catch(err) {
-      callback(err)
+      var keys, seedBytesUint8Array, boxKeyPairSeed,
+        signKeyPairSeed, signKeyPair
+
+      // Convert scrypt Array of Bytes to Uint8Array
+      seedBytesUint8Array = new Uint8Array(scryptByteArray)
+
+      // First 32 Bytes of scrypt seed for encryption keys
+      // Note : first 32 Bytes are the same for dkLen 32 (old way) and 64!
```
Comment (elimisteve) [inline index.js:63]: This fact also seems to suggest that backward-compatibility was maintained, which is fantastic. /cc @jo @grempe

File: test.js (hunk 6 above)
Comment (elimisteve) [reply test.js:12]: Ah, one just needs to take the `Uint8Array` representation of the public key and then base58-encode it to get the "miniLock ID", it appears, but the derived key pair should be identical, methinks. /cc @jo

**Comment by grempe** [comment]
Sorry for the delay in responding. I didn't test for compatibility with Minilock. But I think if the original version of this lib was compatible, then the PR I added also would be. I don't think the original version of this lib made any claims one way or the other about being compatible.

I think adding several tests that compare a key generated by minilock and session25519 would be good to add to this lib prior to relying on it and to ensure that any such compatibility doesn't break in the future.

**Comment by elimisteve** [comment]
Hi @grempe, you're right; adding tests is important and I plan to add them as I get closer to relying on said compatibility. Thanks!
